src/vm/amd64/PInvokeStubs.asm

```diff
@@ -135,4 +143,103 @@ NESTED_ENTRY VarargPInvokeGenILStub, _TEXT
+;
+;
+LEAF_ENTRY JIT_PInvokeBegin, _TEXT
+        
+        mov             rax, qword ptr [s_gsCookie]
+        mov             qword ptr [rcx], rax
+        add             rcx, SIZEOF_GSCookie
+
+        ;; set first slot to the value of InlinedCallFrame::`vftable' (checked by runtime code)
+        lea             rax,[??_7InlinedCallFrame@@6B@]
+        mov             qword ptr [rcx], rax
+
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_Datum], 0
+
+        mov             rax, rsp
+        add             rax, 8
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCallSiteSP], rax
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCalleeSavedFP], rbp
+
+        mov             rax, [rsp]
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCallerReturnAddress], rax
+
+        INLINE_GETTHREAD rax 
+        ;; pFrame->m_Next = pThread->m_pFrame;
+        mov             rdx, qword ptr [rax + OFFSETOF__Thread__m_pFrame]
+        mov             qword ptr [rcx + OFFSETOF__Frame__m_Next], rdx
+
+        ;; pThread->m_pFrame = pFrame;
+        mov             qword ptr [rax + OFFSETOF__Thread__m_pFrame], rcx
+
+        ;; pThread->m_fPreemptiveGCDisabled = 0
+        mov             dword ptr [rax + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 0
+
+        ret
+
+LEAF_END JIT_PInvokeBegin, _TEXT
+
+;
+; in:
+; InlinedCallFrame (rcx) = pointer to the InlinedCallFrame data, including the GS cookie slot (GS cookie right 
+;                          before actual InlinedCallFrame data)
+;
+;
+LEAF_ENTRY JIT_PInvokeEnd, _TEXT
+
+        add             rcx, SIZEOF_GSCookie
+
+        INLINE_GETTHREAD rdx 
+
+        ;; rcx = pFrame
+        ;; rdx = pThread
+
+        ;; pThread->m_fPreemptiveGCDisabled = 1
+        mov             dword ptr [rdx + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 1
+
+        ;; Check return trap
+        cmp             [g_TrapReturningThreads], 0
+        jnz             JIT_PInvokeEndRarePath
+
+        ;; pThread->m_pFrame = pFrame->m_Next
+        mov             rax, qword ptr [rcx + OFFSETOF__Frame__m_Next]
+        mov             qword ptr [rdx + OFFSETOF__Thread__m_pFrame], rax
+
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCallerReturnAddress], 0
+
+        ret
+
+LEAF_END JIT_PInvokeEnd, _TEXT
+
+;
+; in:
+; InlinedCallFrame (rcx) = pointer to the InlinedCallFrame data
+; Thread           (rdx) = pointer to current thread
+;
+;
+NESTED_ENTRY JIT_PInvokeEndRarePath, _TEXT
+
+        alloc_stack         MIN_SIZE
+        save_reg_postrsp    r14, MIN_SIZE + 08h
+        END_PROLOGUE
+
+        ;; Save frame in callee saved registers
+        mov             r14, rcx
+
+        ;; Call helper
+        call            JIT_RareDisableAndPopFrameFromThreadHelper
+
+        mov             qword ptr [r14 + OFFSETOF__InlinedCallFrame__m_pCallerReturnAddress], 0
```

Comment:
You can move this bit into the C++ helper as well, and make the C++ helper itself to be `JIT_PInvokeEndRarePath`

Reply:
E.g. `JIT_PInvokeEnd` would be:

```
       ;; Check return trap
        cmp             [g_TrapReturningThreads], 0
        jnz             RarePath

        ;; pThread->m_pFrame = pFrame->m_Next
        mov             rax, qword ptr [rcx + OFFSETOF__Frame__m_Next]
        mov             qword ptr [rdx + OFFSETOF__Thread__m_pFrame], rax

        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCallerReturnAddress], 0

        ret

RarePath:
        jmp             JIT_PInvokeEndRarePath

LEAF_END JIT_PInvokeEnd, _TEXT
```

Reply:
Thanks for the suggestion Jan. It makes much more sense to do it that way :)